lib/services/usagereporter.go

```diff
@@ -0,0 +1,202 @@
+	return &UsageUIOnboardCompleteGoToDashboardClickEvent{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// UsageUIOnboardAddFirstResourceClickEvent is a UI event sent when a user
+// clicks the "add first resource" button.
+type UsageUIOnboardAddFirstResourceClickEvent prehogv1.UIOnboardAddFirstResourceClickEvent
+
+func (u *UsageUIOnboardAddFirstResourceClickEvent) Anonymize(a utils.Anonymizer) UsageAnonymizable {
+	return &UsageUIOnboardAddFirstResourceClickEvent{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// UsageUIOnboardAddFirstResourceLaterClickEvent is a UI event sent when a user
+// clicks the "add first resource later" button.
+type UsageUIOnboardAddFirstResourceLaterClickEvent prehogv1.UIOnboardAddFirstResourceLaterClickEvent
+
+func (u *UsageUIOnboardAddFirstResourceLaterClickEvent) Anonymize(a utils.Anonymizer) UsageAnonymizable {
+	return &UsageUIOnboardAddFirstResourceLaterClickEvent{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// UsageUIOnboardSetCredentialSubmit is an UI event sent during registration
+// when the user configures login credentials.
+type UsageUIOnboardSetCredentialSubmit prehogv1.UIOnboardSetCredentialSubmitEvent
+
+func (u *UsageUIOnboardSetCredentialSubmit) Anonymize(a utils.Anonymizer) UsageAnonymizable {
+	return &UsageUIOnboardSetCredentialSubmit{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// UsageUIOnboardRegisterChallengeSubmit is a UI event sent during registration
+// when the MFA challenge is completed.
+type UsageUIOnboardRegisterChallengeSubmit prehogv1.UIOnboardRegisterChallengeSubmitEvent
+
+func (u *UsageUIOnboardRegisterChallengeSubmit) Anonymize(a utils.Anonymizer) UsageAnonymizable {
+	return &UsageUIOnboardRegisterChallengeSubmit{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// UsageUIOnboardRecoveryCodesContinueClick is a UI event sent during initial
+// tenant onboarding when a user configures recovery codes.
+type UsageUIOnboardRecoveryCodesContinueClick prehogv1.UIOnboardRecoveryCodesContinueClickEvent
+
+func (u *UsageUIOnboardRecoveryCodesContinueClick) Anonymize(a utils.Anonymizer) UsageAnonymizable {
+	return &UsageUIOnboardRecoveryCodesContinueClick{
+		UserName: a.Anonymize([]byte(u.UserName)),
+	}
+}
+
+// ConvertUsageEvent converts a usage event from an API object into an
+// anonymizable event. All events that can be submitted externally via the Auth
+// API need to be defined here.
+func ConvertUsageEvent(event *usageevents.UsageEventOneOf, username string) (UsageAnonymizable, error) {
+	switch e := event.GetEvent().(type) {
+	case *usageevents.UsageEventOneOf_UiBannerClick:
+		return &UsageUIBannerClick{
+			UserName: username,
+			Alert:    e.UiBannerClick.Alert,
+		}, nil
+	case *usageevents.UsageEventOneOf_UiOnboardGetStartedClick:
+		return &UsageUIOnboardGetStartedClickEvent{
+			UserName: e.UiOnboardGetStartedClick.Username,
```

Comment:
Annoyingly the username is always set, if coming from the proxy client we get the proxy's own username. The rule is just to pass through usernames for the few events that embed them, otherwise use the identity username.

I'll add a comment to explain that.

Reply:
I think we probably could, yeah, but I'm not totally convinced of the need to do so. I figure we can change behavior here if it comes up in the future, but can keep the implementation simple for now?